src/algorithms/sdp.jl

```diff
@@ -0,0 +1,161 @@
+# Semidefinite programming
+
+# Types
+MatTypeA = Union{Matrix}
+TypeA = Union{Real, Matrix}
+
+# Util functions
+function dof_to_dim(dof::Int)
+    @assert is_triangular(dof) "dof must be a triangular number. "
+    return trunc(Int, (sqrt(8*dof+1)-1)÷2)
+end
+
+function dof_to_length(dof::Int)
+    dim = dof_to_dim(dof)
+    return dim^2
+end
+
+dim_to_dof(dim::Int) = (dim^2+dim)÷2
+
+function apply_sd_constraint(A0::TypeA, Ai::AbstractVector, x::AbstractVector)
+    return reduce(.+, Ai.*x).+A0
```

Comment:
We shouldn't assume the LMI constraint, just a single matrix-valued function `A(x)` that should be PSD. Let's discuss this tomorrow.

Reply:
Another option is we can just input a single `A` that sums up `A0` and all `Ai`, which represents an element-wise linear constraint on the objective matrix. But if one constraint is related to x0 but not a scalar, we'll have to make it something like an expression `:(x[10]+2*x[11]+2)`

Reply:
It's a generalized version of regular inequality constraint. So when using this we don't need other `ineq_constraints` to avoid ambiguity.

Reply:
Oh I got another idea. We can actually minimally only have `A0`, and for other constraints just use regular `add_ineq_constraint!`

Reply:
We can combine `A0` and `Ai` to specify constraints on `M`. Then use regular `InEqconstraint` for other constraints 
 on `x`. This should be as expressive as current design.